File: src/main/java/brs/feesuggestions/FeeSuggestionCalculator.java
package brs.feesuggestions;

import brs.Block;
import brs.BlockchainProcessor;
import brs.BlockchainProcessor.Event;
import brs.Constants;
import brs.DependencyProvider;
import brs.db.store.BlockchainStore;

import java.util.Arrays;

import static brs.Constants.FEE_QUANT;

public class FeeSuggestionCalculator { // TODO interface

  // index 0 = oldest, length-1 = newest
  private final Block[] latestBlocks;
  private final DependencyProvider dp;
  private FeeSuggestion feeSuggestion = new FeeSuggestion(FEE_QUANT, FEE_QUANT, FEE_QUANT);

  public FeeSuggestionCalculator(DependencyProvider dp) {
    this.dp = dp;
    latestBlocks = new Block[Constants.FEE_SUGGESTION_MAX_HISTORY_LENGTH];
    dp.blockchainProcessor.addListener(this::newBlockApplied, Event.AFTER_BLOCK_APPLY);
  }

  public FeeSuggestion giveFeeSuggestion() {
    if (latestBlocksIsEmpty()) {
      fillInitialHistory();
      recalculateSuggestion();
    }

    return feeSuggestion;
  }

  private void newBlockApplied(Block block) {
    if (latestBlocksIsEmpty()) {
      fillInitialHistory();
    }

    pushNewBlock(block);
    recalculateSuggestion();
  }

  private void fillInitialHistory() {
    dp.blockchainStore.getLatestBlocks(latestBlocks.length).forEach(this::pushNewBlock);
  }

  private boolean latestBlocksIsEmpty() {
    for (Block latestBlock : latestBlocks) {
      if (latestBlock == null) return true;
    }
    return false;
  }

  private void pushNewBlock(Block block) {
    if (block == null) return;
    // Skip index 0 as we want to remove this one
    if (latestBlocks.length - 1 >= 0) System.arraycopy(latestBlocks, 1, latestBlocks, 0, latestBlocks.length - 1);
    latestBlocks[latestBlocks.length - 1] = block;
  }

  private void recalculateSuggestion() {
    try {
      int lowestAmountTransactionsNearHistory = Arrays.stream(latestBlocks).mapToInt(b -> b.getTransactions().size()).min().orElse(1);
      int averageAmountTransactionsNearHistory = (int) Math.ceil(Arrays.stream(latestBlocks).mapToInt(b -> b.getTransactions().size()).average().orElse(1));
      int highestAmountTransactionsNearHistory = Arrays.stream(latestBlocks).mapToInt(b -> b.getTransactions().size()).max().orElse(1);

      long cheapFee = (1 + lowestAmountTransactionsNearHistory) * FEE_QUANT;
      long standardFee = (1 + averageAmountTransactionsNearHistory) * FEE_QUANT;
      long priorityFee = (1 + highestAmountTransactionsNearHistory) * FEE_QUANT;

      feeSuggestion = new FeeSuggestion(cheapFee, standardFee, priorityFee);
    } catch (NullPointerException ignored) { // Can happen if there are less than latestBlocks.length blocks in the blockchain
    }
  }
}

File: src/main/java/brs/feesuggestions/FeeSuggestionCalculator.kt
package brs.feesuggestions

import brs.Block
import brs.BlockchainProcessor.Event
import brs.Constants
import brs.Constants.FEE_QUANT
import brs.DependencyProvider
import kotlinx.coroutines.runBlocking
import kotlin.math.ceil

class FeeSuggestionCalculator(private val dp: DependencyProvider, maxHistoryLength: Int = Constants.FEE_SUGGESTION_MAX_HISTORY_LENGTH) { // TODO interface
    // index 0 = oldest, length-1 = newest
    private val latestBlocks = arrayOfNulls<Block>(maxHistoryLength)
    private var feeSuggestion = FeeSuggestion(FEE_QUANT, FEE_QUANT, FEE_QUANT)

    init {
        runBlocking {
            dp.blockchainProcessor.addListener(Event.AFTER_BLOCK_APPLY) { newBlockApplied(it) }
        }
    }

    suspend fun giveFeeSuggestion(): FeeSuggestion {
        if (latestBlocksIsEmpty()) {
            fillInitialHistory()
            recalculateSuggestion()
        }

        return feeSuggestion
    }

    private suspend fun newBlockApplied(block: Block) {
        if (latestBlocksIsEmpty()) {
            fillInitialHistory()
        }

        pushNewBlock(block)
        recalculateSuggestion()
    }

    private suspend fun fillInitialHistory() {
        dp.blockchainStore.getLatestBlocks(latestBlocks.size).forEach { this.pushNewBlock(it) }
    }

    private fun latestBlocksIsEmpty(): Boolean {
        for (latestBlock in latestBlocks) {
            if (latestBlock == null) return true
        }
        return false
    }

    private fun pushNewBlock(block: Block?) {
        if (block == null) return
        // Skip index 0 as we want to remove this one
        if (latestBlocks.size - 1 >= 0) System.arraycopy(latestBlocks, 1, latestBlocks, 0, latestBlocks.size - 1)
        latestBlocks[latestBlocks.size - 1] = block
    }

    private suspend fun recalculateSuggestion() {
        try {
            val transactionSizes = latestBlocks
                    .filterNotNull()
                    .map { it.getTransactions().size }
            val lowestAmountTransactionsNearHistory = transactionSizes.min() ?: 1
            val averageAmountTransactionsNearHistory = ceil(transactionSizes.average().let { return@let if (it.isNaN()) 1.0 else it }).toLong()
            val highestAmountTransactionsNearHistory = transactionSizes.max() ?: 1

            val cheapFee = (1 + lowestAmountTransactionsNearHistory) * FEE_QUANT
            val standardFee = (1 + averageAmountTransactionsNearHistory) * FEE_QUANT
            val priorityFee = (1 + highestAmountTransactionsNearHistory) * FEE_QUANT

            feeSuggestion = FeeSuggestion(cheapFee, standardFee, priorityFee)
        } catch (ignored: NullPointerException) { // Can happen if there are less than latestBlocks.length blocks in the blockchain
        }

    }
}
